Sources/WeaverDI/ThreadSafeBuilderStoreDecorator.swift

```diff
@@ -0,0 +1,44 @@
+//
+//  ThreadSafeBuilderStoringDecorator.swift
+//  WeaverDIPackageDescription
+//
+//  Created by Bruno Mazzo on 5/31/18.
+//
+
+import Foundation
+
+final class ThreadSafeBuilderStoreDecorator: BuilderStoring {
+    
+    private let builders: BuilderStoring
+    private let queue: DispatchQueue = DispatchQueue(label: "\(ThreadSafeBuilderStoreDecorator.self)")
+    
+    init(builders: BuilderStoring) {
+        self.builders = builders
+    }
+    
+    func get<B>(for key: InstanceKey, isCalledFromAChild: Bool) -> B? {
+        var getValue: B? = nil
+        
+        queue.sync {
+            getValue = self.builders.get(for: key, isCalledFromAChild: isCalledFromAChild)
+        }
+        
+        return getValue
+    }
+    
+    func set<B>(builder: B, scope: Scope, for key: InstanceKey) {
+        self.queue.async {
```

Comment:
💯 

I thought the default `DispatchQueue` was concurrent, but you're actually right it's serial, so the barrier is not useful here. 

Actually, even with a concurrent, the right place to put the barrier would be on the `set`, but since it's `sync`, the barrier would lead to a deadlock.

I think making that queue concurrent would be more efficient, but I can't see how to achieve that. Any suggestion on how to do that?

Reply:
In any case, a serial is probably good enough for now